AIX platform build fix

gopukrishnasIBM opened the pull request:
These file changes fix the build issues on the IBM AIX platform. 
Since AIX is not taken care of in the existing library, "BSON_USE_LEGACY_GCC_ATOMICS" is not getting defined, and "DEF_ATOMIC_OP" macro which dynamically creates code for atomic functions based on the Operating system, generates incorrect code with multiple return statements (Code expanded using xlC -E ). 
```
static __inline__ signed char bson_atomic_int8_fetch_add(signed char volatile * a, signed char addend, enum bson_memory_order ord) {
   do {
      switch (ord) {
      case bson_memory_order_acq_rel:
         BSON_IF_MSVC(
            return _InterlockedExchangeAdd8(a, addend);) return __tg_builtin("__sync_fetch_and_add", a, addend);
      case bson_memory_order_seq_cst:
         BSON_IF_MSVC(
            return _InterlockedExchangeAdd8(a, addend);) return __tg_builtin("__sync_fetch_and_add", a, addend);
      case bson_memory_order_acquire:
         BSON_IF_MSVC(
            return _InterlockedExchangeAdd8(a, addend);) return __tg_builtin("__sync_fetch_and_add", a, addend);
      case bson_memory_order_consume:
         BSON_IF_MSVC(
            return _InterlockedExchangeAdd8(a, addend);) return __tg_builtin("__sync_fetch_and_add", a, addend);
      case bson_memory_order_release:
         BSON_IF_MSVC(
            return _InterlockedExchangeAdd8(a, addend);) return __tg_builtin("__sync_fetch_and_add", a, addend);
      case bson_memory_order_relaxed:
         BSON_IF_MSVC(
            return _InterlockedExchangeAdd8(a, addend);) return __tg_builtin("__sync_fetch_and_add", a, addend);
      default:
         do {
            fprintf(( & _iob[2]), "%s:%d %s(): Unreachable code reached: %s\n", "bson-atomic.h", 361, __func__, "Invalid bson_memory_order value");
            abort();
         } while (0);
      }
   } while (0);
}
``` 
After the fix, "DEF_ATOMIC_OP" macro is generating legacy atomic functions on the AIX platform, which solves the build issues. 
```
static __inline__ signed char bson_atomic_int8_fetch_add(signed char volatile * a, signed char addend, enum bson_memory_order ord) {
   do {
      switch (ord) {
      case bson_memory_order_acq_rel:
         return __tg_builtin("__sync_fetch_and_add", a, addend);
      case bson_memory_order_seq_cst:
         return __tg_builtin("__sync_fetch_and_add", a, addend);
      case bson_memory_order_acquire:
         return __tg_builtin("__sync_fetch_and_add", a, addend);
      case bson_memory_order_consume:
         return __tg_builtin("__sync_fetch_and_add", a, addend);
      case bson_memory_order_release:
         return __tg_builtin("__sync_fetch_and_add", a, addend);
      case bson_memory_order_relaxed:
         return __tg_builtin("__sync_fetch_and_add", a, addend);
      default:
         do {
            fprintf(( & _iob[2]), "%s:%d %s(): Unreachable code reached: %s\n", "bson-atomic.h", 363, __func__, "Invalid bson_memory_order value");
            abort();
         } while (0);
      }
   } while (0);
}
```

vector-of-bool commented on src/libbson/src/bson/bson-compat.h:
```diff
@@ -332,6 +332,13 @@ typedef signed char bool;
 #define BSON_IF_MSVC(...)
 /** Expands the arguments if compiling with GCC or Clang, otherwise empty */
 #define BSON_IF_GNU_LIKE(...) __VA_ARGS__
+#elif defined(_AIX)
+/** Expands the arguments if compiling with MSVC, otherwise empty */
+#define BSON_IF_MSVC(...)
+/** Expands the arguments if compiling with GCC or Clang, otherwise empty */
+#define BSON_IF_GNU_LIKE(...) 
+/** Expands the arguments if compiling with AIX, otherwise empty */
+#define BSON_IF_GNU_LEGACY_ATOMICS(...) __VA_ARGS__
```
The `BSON_IF_GNU_LEGACY_ATOMICS` is only used within `bson-atomic.h`, and is already defined if `BSON_USE_LEGACY_GCC_ATOMICS` is defined, so it need not be defined here.

If I understand correctly: If you are using a recent version of xlC, the branch for `defined(__clang__)` should be taken on line 330. If you are using a legacy version, a `|| defined(__xlC__)` may need to be added on line 330. This should add the necessary `#define BSON_IF_MSVC` that will suppress its appearance in `bson-atomic.h`.

vector-of-bool reviewed (commented):
Thank you for bringing this to our attention!

I've left a remark in the addition in `bson-compat.h` with a possible simpler alternative fix.

gopukrishnasIBM commented:
@vector-of-bool based on your comment, I reverted the changes in bson-compat.h.  

On bson-atomic.h, currently BSON_USE_LEGACY_GCC_ATOMICS logic is like
```
#ifdef BSON_USE_LEGACY_GCC_ATOMICS
#undef BSON_IF_GNU_LIKE
#define BSON_IF_GNU_LIKE(...)
#define BSON_IF_GNU_LEGACY_ATOMICS(...) __VA_ARGS__
```
Since we are not taking care of the "BSON_IF_MSVC(...)" macro, it also expands along with " BSON_IF_GNU_LEGACY_ATOMICS(...)" macro, resulting in code 
```
 case bson_memory_order_acq_rel:
         BSON_IF_MSVC(
            return _InterlockedExchangeAdd8(a, addend);) return __tg_builtin("__sync_fetch_and_add", a, addend);
```

So, I changed BSON_USE_LEGACY_GCC_ATOMICS logic to 
```
#ifdef BSON_USE_LEGACY_GCC_ATOMICS
#undef BSON_IF_GNU_LIKE
#define BSON_IF_GNU_LIKE(...)
#define BSON_IF_MSVC(...)
#define BSON_IF_GNU_LEGACY_ATOMICS(...) __VA_ARGS__
```
and now the resulting code is
```
case bson_memory_order_acq_rel:
         return __tg_builtin("__sync_fetch_and_add", a, addend);
```

vector-of-bool commented:
Thank you! I'll let CI run and see if there are any unexpected issues appear.

vector-of-bool commented:
I apologize for the delay, thank you for your patience.

The existing task failures are known unrelated issues. These changes LGTM. Thank you!

gopukrishnasIBM commented:
Thank you @vector-of-bool
